text/3750-cfg-os-version-min.md

```diff
@@ -0,0 +1,113 @@
+For `x86_64-win7-pc-windows-msvc` the minimum API version is "6.1.7600" which corresponds to Windows 7.
+However, inferring the API version from the target name isn't ideal especially as it can change over time.
+
+Instead you use the `os_version_min` predicates to specify the minimum API levels of various parts of the operating system.  For example:
+
+* `os_version_min(“windows”, <string>)` would test the [minimum build version](https://gaijin.at/en/infos/windows-version-numbers) of Windows.
+* `os_version_min(“libc”, <string>)` would test the version of libc.
+* `os_version_min(“kernel”, <string>)` would test the version of the kernel.
+
+Let’s use `os_version_min(“windows”, …)` as an example.  It should be clear how this example would be extended to the other `cfg` predicates. The predicate allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(os_version_min(“windows”, "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
+```
+
+For targets where `os_version_min(“windows”, …)` does not make sense (i.e., non-Windows targets), the `cfg` predicate will return `false` and emit a warning saying that the particular `cfg` predicate is not supported on that target. Therefore, it's important to pair `os_version_min(“windows”, …)` with a `cfg(windows)` using the existing mechanisms for combining `cfg` predicates.
+
+The above example works exactly the same way with the other platform API `cfg` predicates just with different values and different target support.
+
+These predicates do not assume any semantic versioning information. The specified predicates are simply listed in order. The only semantics that are assumed is that code compiled with the `cfg` predicates works for all versions greater than or equal to that version.
+
+**Note:** Here it would be important to link to documentation showing the `cfg` predicates and the different version strings that are supported.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `os_version_min` predicate allows users to conditionally compile code based on the API version supported by the target platform.
+Each platform is responsible for defining a default key, a set of keys it supports, and functions that are able to compare the version strings they use.
+A set of comparison functions can be provided by `rustc` for common formats such as 2- and 3-part semantic versioning.
+When a platform detects a key it doesn’t support it will return `false` and emit a warning.
+
+Each target platform will set the minimum API versions it supports.
```

Comment:
So there is no way for the user to configure which minimum API version they want to use? That would make it impossible for eg the libc crate to gate api's behind `#[cfg(os_version_min)]` corresponding to the libc version that introduced the API in question.

Reply:
Yes. One of the issues with the original proposals was there was too much going on. This RFC aims to add the minimally useful feature. Adding compiler flags and Cargo configs to control the minimum version can be a future extension.

With this RFC it is still possible for libc to gate APIs like that. However, you'd need a new target in order to set a different minimal libc version.